pkg/cache/priority_queue.go

```diff
@@ -0,0 +1,191 @@
+		return nil
+	}
+	delete(pq.items, entry.Value)
+	return entry
+}
+
+// Peek return the highest priority element
+func (pq *PriorityQueue) Peek() *Entry {
+	return pq.queue.Peek()
+}
+
+// Update update the priority of the given element
+func (pq *PriorityQueue) Update(old *Entry, priority int) {
+	pq.queue.update(old, old.Value, priority)
+}
+
+// GetAll return limited Entries, if limit = -1 ,it will return all entries
+func (pq *PriorityQueue) GetAll() []*Entry {
+	result := pq.queue.GetAll()
+	return result
+}
+
+// RemoveValue remove element
+func (pq *PriorityQueue) RemoveValue(value interface{}) {
+	if entry, ok := pq.items[value]; ok {
+		pq.queue.remove(entry.Index)
+		delete(pq.items, value)
+	}
+}
+
+// RemoveValues remove elements
+func (pq *PriorityQueue) RemoveValues(values []interface{}) {
+	for _, v := range values {
+		pq.RemoveValue(v)
+	}
+}
+
+// Has it will return true if queue has the value
+func (pq *PriorityQueue) Has(v interface{}) bool {
+	if value, ok := pq.items[v]; ok && value.Index != -1 {
+		return true
+	}
+	return false
+}
+
+// Size return the size of queue
+func (pq *PriorityQueue) Size() int {
+	return pq.queue.Len()
+}
+
+// Entry internal struct for record element
+type Entry struct {
+	Value    interface{}
+	Priority int
+	Index    int
+}
+
+// priorityHeap implements heap.Interface and holds entries.
+type priorityHeap []*Entry
+
+// Len
+func (pq priorityHeap) Len() int { return len(pq) }
+
+// Less
+func (pq priorityHeap) Less(i, j int) bool {
+	return pq[i].Priority >= pq[j].Priority
+}
+
+// Swap
+func (pq priorityHeap) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].Index = i
+	pq[j].Index = j
+}
+
+// Push push element to heap
+func (pq *priorityHeap) Push(x interface{}) {
+	entry := x.(*Entry)
+	entry.Index = len(*pq)
+	*pq = append(*pq, entry)
+	pq.fix(entry)
+}
+
+// Pop pop the highest Priority element from heap
+func (pq *priorityHeap) Pop() interface{} {
+	return pq.remove(0)
+}
+
+// Peek peek the highest priority element from heap
+func (pq *priorityHeap) Peek() *Entry {
+	if pq.Len() <= 0 {
+		return nil
+	}
+	old := *pq
+	return old[0]
+}
+
+// GetAll return the limited size of elements
+func (pq *priorityHeap) GetAll() []*Entry {
+	tmp := *pq
+	return tmp[:]
+}
+
+// remove remove index from heap
+// it will swap the lowest priority element of the target ,and then fix the index
+func (pq *priorityHeap) remove(id int) interface{} {
+	if id >= pq.Len() {
+		return nil
+	}
+	old := *pq
+	result := old[id]
+	pq.Swap(id, len(old)-1)
+	*pq = old[:len(old)-1]
+	// skip if element is the last element
+	if id < pq.Len() {
+		pq.fix(old[id])
+	}
+	result.Index = -1 // safe mark
+	return result
+}
+
+// fix the entry index
+func (pq *priorityHeap) fix(entry *Entry) {
+	if pq.Len() <= entry.Index {
+		return
+	}
+	heap.Fix(pq, entry.Index)
+}
+
+// update modifies the priority and value of an entry in the queue
+func (pq *priorityHeap) update(entry *Entry, value interface{}, priority int) {
```

Comment:
```suggestion
func (pq *priorityHeap) update(entry *Entry, priority int) {
```

Changing the `value` also requires changing the map at the same time. We don't have this need.